File: threedollar-api/src/main/java/com/depromeet/threedollar/api/controller/auth/AuthController.java
package com.depromeet.threedollar.api.controller.auth;

import com.depromeet.threedollar.api.config.interceptor.Auth;
import com.depromeet.threedollar.api.config.resolver.UserId;
import com.depromeet.threedollar.api.common.dto.ApiResponse;
import com.depromeet.threedollar.api.service.auth.AuthService;
import com.depromeet.threedollar.api.service.auth.dto.request.LoginRequest;
import com.depromeet.threedollar.api.service.auth.dto.request.SignOutRequest;
import com.depromeet.threedollar.api.service.auth.dto.request.SignUpRequest;
import com.depromeet.threedollar.api.service.auth.dto.response.LoginResponse;
import com.depromeet.threedollar.api.service.token.TokenService;
import com.depromeet.threedollar.api.service.token.dto.UserTokenDto;
import com.depromeet.threedollar.common.exception.ValidationException;
import com.depromeet.threedollar.domain.domain.user.UserSocialType;
import io.swagger.annotations.ApiImplicitParam;
import io.swagger.annotations.ApiOperation;
import lombok.RequiredArgsConstructor;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RestController;

import javax.validation.Valid;

@RequiredArgsConstructor
@RestController
public class AuthController {

    private final AuthService appleAuthService;
    private final AuthService kaKaoAuthService;
    private final TokenService jwtService;

    @ApiOperation("회원가입을 요청합니다")
    @PostMapping("/api/v2/signup")
    public ApiResponse<LoginResponse> signUp(@Valid @RequestBody SignUpRequest request) {
        Long userId = signUpBySocialType(request);
        return ApiResponse.success(LoginResponse.of(jwtService.encode(new UserTokenDto(userId))));
    }

    private Long signUpBySocialType(SignUpRequest request) {
        if (request.getSocialType().equals(UserSocialType.KAKAO)) {
            return kaKaoAuthService.signUp(request);
        }
        if (request.getSocialType().equals(UserSocialType.APPLE)) {
            return appleAuthService.signUp(request);
        }
        throw new ValidationException(String.format("허용하지 않는 소셜 타입 (%s) 입니다.", request.getSocialType()));
    }

    @ApiOperation("로그인을 요청합니다")
    @PostMapping("/api/v2/login")
    public ApiResponse<LoginResponse> login(@Valid @RequestBody LoginRequest request) {
        UserTokenDto tokenDto = new UserTokenDto(loginBySocialType(request));
        return ApiResponse.success(LoginResponse.of(jwtService.encode(tokenDto)));
    }

    private Long loginBySocialType(LoginRequest request) {
        if (request.getSocialType().equals(UserSocialType.KAKAO)) {
            return kaKaoAuthService.login(request);
        }
        if (request.getSocialType().equals(UserSocialType.APPLE)) {
            return appleAuthService.login(request);
        }
        throw new ValidationException(String.format("허용하지 않는 소셜 타입 (%s) 입니다.", request.getSocialType()));
    }

    @ApiOperation("[인증] 회원탈퇴를 요청합니다")
    @ApiImplicitParam(name = "Authorization", value = "Access Token", required = true, paramType = "header")
    @Auth
    @DeleteMapping("/api/v2/signout")
    public ApiResponse<String> signOut(@Valid SignOutRequest request, @UserId Long userId) {
        signOutBySocialType(request, userId);
        return ApiResponse.SUCCESS;
    }

    private void signOutBySocialType(SignOutRequest request, Long userId) {
        if (request.getSocialType().equals(UserSocialType.KAKAO)) {
            kaKaoAuthService.signOut(userId);
        }
        if (request.getSocialType().equals(UserSocialType.APPLE)) {
            appleAuthService.signOut(userId);
        }
        throw new ValidationException(String.format("허용하지 않는 소셜 타입 (%s) 입니다.", request.getSocialType()));
    }

}

File: threedollar-api-boss/src/main/kotlin/com/depromeet/threedollar/boss/api/controller/auth/AuthController.kt
package com.depromeet.threedollar.boss.api.controller.auth

import com.depromeet.threedollar.application.common.dto.ApiResponse
import com.depromeet.threedollar.boss.api.config.interceptor.Auth
import com.depromeet.threedollar.boss.api.config.session.SessionConstants.BOSS_ACCOUNT_ID
import com.depromeet.threedollar.boss.api.service.auth.AuthServiceProvider
import com.depromeet.threedollar.boss.api.service.auth.dto.request.LoginRequest
import com.depromeet.threedollar.boss.api.service.auth.dto.response.LoginResponse
import com.depromeet.threedollar.boss.api.service.auth.SignupService
import com.depromeet.threedollar.boss.api.service.auth.dto.request.SignupRequest
import io.swagger.annotations.ApiOperation
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RestController
import javax.servlet.http.HttpSession
import javax.validation.Valid

@RestController
class AuthController(
    private val httpSession: HttpSession,
    private val authServiceProvider: AuthServiceProvider,
    private val signupService: SignupService
) {

    @ApiOperation("사장님 계정의 회원가입을 요청합니다. (차후 승인이 필요합니다)")
    @PostMapping("/v1/signup")
    fun signup(
        @Valid @RequestBody request: SignupRequest
    ): ApiResponse<String> {
        val authService = authServiceProvider.getAuthService(request.socialType)
        val socialId = authService.getSocialId(LoginRequest(request.token, request.socialType))
        signupService.signup(request, socialId)
        return ApiResponse.SUCCESS
    }

    @ApiOperation("사장님 계정으로 로그인을 요청합니다.")
    @PostMapping("/v1/login")
    fun login(
        @Valid @RequestBody request: LoginRequest
    ): ApiResponse<LoginResponse> {
        val authService = authServiceProvider.getAuthService(request.socialType)
        val accountId = authService.login(request)
        httpSession.setAttribute(BOSS_ACCOUNT_ID, accountId)
        return ApiResponse.success(LoginResponse(httpSession.id, accountId))
    }

    @ApiOperation("[인증] 사장님 계정으로 로그아웃을 요청합니다.")
    @Auth
    @PostMapping("/v1/logout")
    fun logout(): ApiResponse<String> {
        httpSession.removeAttribute(BOSS_ACCOUNT_ID)
        return ApiResponse.SUCCESS
    }

}
